### ISSUE TITLE
MSI configuration issues when using provider block

### ISSUE THREAD

**Issue by adamday2** [hashicorp/terraform-provider-azurerm#3035]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.11.13
+ provider.azurerm v1.23.0

### Affected Resource(s)

* `azurerm` `provider` block
* MSI logins 

### Azure MSI Login Result

```sh
$ az login --identity
[
  {
    "environmentName": "AzureUSGovernment",
    "id": "[redacted]",
    "isDefault": true,
    "name": "[redacted]",
    "state": "Enabled",
    "tenantId": "[redacted]",
    "user": {
      "assignedIdentityInfo": "MSI",
      "name": "systemAssignedIdentity",
      "type": "servicePrincipal"
    }
  }
]
```

#### Test 1 - *Failure*
```hcl
provider "azurerm" {
   use_msi = true
}
```
Result:

It appears the `subscription_id` and `environment` are not being picked up appropriately from the MSI.
```sh
Error: Error refreshing state: 1 error(s) occurred:

* provider.azurerm: Unable to list provider registration status, it is possible that this is due to invalid credentials or the service principal does not have permission to use the Resource Manager API, Azure error: resources.ProvidersClient#List: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="SubscriptionNotFound" Message="The subscription 'providers' could not be found."
```
Additional relevant DEBUG output showing no subscription ID and the incorrect management endpoint for the environment.
```
AzureRM Request:
GET /subscriptions//providers?api-version=2016-02-01 HTTP/1.1
Host: management.azure.com
```

#### Test 2 - *Failure*

Attempting to manually configure the `environment` and `subscription_id`
```hcl
provider "azurerm" {
   use_msi = true
   subscription_id = "[redacted]"
   environment = "usgovernment"
}
```
Result:
```sh
* module.linux_server.provider.azurerm: Error converting Authorization Token to an ADAL Token: Error parsing Token Expiration Date "N/A": Error parsing expiration date "N/A".

CloudShell Error:
parsing time "N/A" as "2006-01-02T15:04:05Z07:00": cannot parse "N/A" as "2006"

CLI Error:
parsing time "N/A" as "2006-01-02 15:04:05.999999": cannot parse "N/A" as "2006"
* module.new_network.provider.azurerm: Error converting Authorization Token to an ADAL Token: Error parsing Token Expiration Date "N/A": Error parsing expiration date "N/A".

CloudShell Error:
parsing time "N/A" as "2006-01-02T15:04:05Z07:00": cannot parse "N/A" as "2006"

CLI Error:
parsing time "N/A" as "2006-01-02 15:04:05.999999": cannot parse "N/A" as "2006"
```
> Note, I have a sneaking suspicion that setting the `subscription_id` is causing `use_msi` to be ignored

#### Test 3 - *Success*

```sh
export ARM_ENVIRONMENT=usgovernment
export ARM_USE_MSI=true
export ARM_SUBSCRIPTION_ID=[redacted]
```

```hcl
provider "azurerm" {
}
```
Result: Success!

### Expected Behavior

* MSI logins should utilize `environment` and `subscription_id` information available from the MSI.
* Provider configuration via the provider block and environment variables should have the same result

### Actual Behavior

* When using MSI, `subscription_id` and `environment` must be manually configured
* When using MSI, `subscription_id` and `environment` can only be configured via environment variables

### Steps to Reproduce

Each result above can be reproduced using the provided configuration and then executing `terraform apply`

### Important Factoids

* Environment: usgovernment

**Comment by raidlman** [comment]
This is still not working in `+ provider.azurerm v2.7.0`.

**Comment by fgarcia-cnb** [comment]
using azurerm v2.14.0, slightly different results, 

if i explicitly set the subscription ID, either in the provider block or as an environment variable, everything works fine. this confirms the managed-identity is working correctly.

i would prefer to not explicitly set the subscription id. but the following doc [https://www.terraform.io/docs/providers/azurerm/guides/managed_service_identity.html](url) mentions "No credentials are stored on the VM, and the only additional information needed to bootstrap the Terraform connection to Azure is the subscription ID and tenant ID."

but the provider block example doesn't show an explicit sub id. and our setup works with just the sub id, and no tenant id. so is the sub id required?

**Comment by ksukenny** [comment]
I'm having issues with MSI in Azure pipelines too. I have a Service Connection setup to use Managed Service Identity on a VMSS that is setup as an agent pool in DevOps and contributor on the Subscription. I've tried several different ways to configure the provider block using some suggestions above. I'm using the Terraform tasks in ADO and it doesn't even seem to be grabbing the Object ID of the MSI since it says the Application Identifier is 'null'.  I'm able to confirm the MSI works with the service connection by running Azure CLI commands in the pipeline, but the terraform tasks aren't using it.

I tried setting the provider version to 2.14 and just putting the subscription_id too, but that didn't work either. I'm using Terraform version 0.13.4.  

Failed to get existing workspaces: Error retrieving keys for Storage Account "storageaccountname": azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/subscription_id/resourceGroups/resource_group_name/providers/Microsoft.Storage/storageAccounts/storageaccount_name/listKeys?api-version=2016-01-01: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"unauthorized_client","error_description":"AADSTS700016: **Application with identifier 'null' was not found in the directory '***'.**
